### PR TITLE
change delete logic

### PR DESCRIPTION
**Pull request by hameizi**
Previous delete logic is write all delete data in eq-delete file although there is same key in pos-delete file.  This PR change this logic to just write the delete data what is not exist in pos-delete file in eq-delete file.  And the old logic will send several delete but one insert of one primary key in one snapshot when we want reperform the data by streaming read, in some scene will cause error (like aggregete count). 
the following is difference between old delete logic and the new. 
table schema: 
```
int key;  (primary key) 
string data;
``` 
**old logic:** 
tx1: 
> insert (1,'aa'); 

tx2: 
>delete (1,'aa');   -->eq-delete file add (1,'aa') 
> insert(1,'bb'); 
>delete (1,'bb');   -->eq-delete file add (1,'bb') pos-delete file add(1,filepath) 
>insert(1,'cc') 

result: 
eq-delete file has (1,'aa'),(1,'bb') 
pos-delete file has (1,filepath) 
 
**new logic:**
tx1: 
> insert (1,'aa'); 

tx2: 
>delete (1,'aa');   -->eq-delete file add (1,'aa') 
> insert(1,'bb'); 
>delete (1,'bb');   --> pos-delete file add(1,filepath) 
>>insert(1,'cc') 

result: 
eq-delete file has (1,'aa') 
pos-delete file has (1,filepath) 
 
Actually the data (1,'bb') is unnecessary in eq-delete file, because when we call function applyPosdelete that (1,'bb') will be delete from result so there is not data match (1,'bb') when we call applyEqdelete.  In streaming read scene the old logic will send two delete (1,'aa'),(1,'bb') of key 1 but one insert (1,'cc') to down opeator, but new logic just send one delete and one insert.

### CODE REVIEW

**Comment by hameizi** [comment]
@openinx @rdblue can you help take a look?

**Comment by rdblue** [comment]
I'm assuming that the equality delete column in your example is the ID. Is that correct?

If so, then there is still some value in writing the delete because you get the previous columns written into the file. That allows reconstructing a CDC stream with the delete.

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java [hunk 1]
```diff
@@ -151,9 +146,14 @@ private void internalPosDelete(StructLike key) {
      * @param row the given row to delete.
      */
     public void delete(T row) throws IOException {
-      internalPosDelete(structProjection.wrap(asStructLike(row)));
+      StructProjection key = structProjection.wrap(asStructLike(row));
+      PathOffset previous = insertedRowMap.get(key);
 
-      eqDeleteWriter.write(row);
+      if (previous != null){
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:152]: Style: there should be a space between `)` and `{`.

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java [hunk 2]
```diff
@@ -120,12 +120,7 @@ public void write(T row) throws IOException {
       // Create a copied key from this row.
       StructLike copiedKey = StructCopy.copy(structProjection.wrap(asStructLike(row)));
 
-      // Adding a pos-delete to replace the old path-offset.
-      PathOffset previous = insertedRowMap.put(copiedKey, pathOffset);
-      if (previous != null) {
-        // TODO attach the previous row if has a positional-delete row schema in appender factory.
-        posDeleteWriter.delete(previous.path, previous.rowOffset, null);
-      }
+      insertedRowMap.put(copiedKey, pathOffset);
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:123]: Why is this changing as well? I think that the logic here was intended to implement upsert just in case there are duplicate inserts without a delete. @openinx, can you take a look at this?

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java [hunk 3]
```diff
@@ -151,9 +146,14 @@ private void internalPosDelete(StructLike key) {
...
-      internalPosDelete(structProjection.wrap(asStructLike(row)));
+      StructProjection key = structProjection.wrap(asStructLike(row));
+      PathOffset previous = insertedRowMap.get(key);
 
-      eqDeleteWriter.write(row);
+      if (previous != null){
+        posDeleteWriter.delete(previous.path, previous.rowOffset, null);
+      } else {
+        eqDeleteWriter.write(row);
+      }
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:156]: Overall the logic change in this function looks reasonable to me.

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java [hunk 4]
```diff
@@ -163,9 +163,14 @@ public void delete(T row) throws IOException {
      * @param key is the projected data whose columns are the same as the equality fields.
      */
     public void deleteKey(T key) throws IOException {
-      internalPosDelete(asStructLike(key));
+      StructLike structLikeKey = asStructLike(key);
+      PathOffset previous = insertedRowMap.get(structLikeKey);
 
-      eqDeleteWriter.write(key);
+      if (previous != null){
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:169]: Style is incorrect here as well.

**Comment by hameizi** [comment]
@rdblue code style fix in new commit.

**Comment by hameizi** [comment]
@rdblue Is there any more question of this PR?

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java (hunk 2 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:124]: @hameizi can you reply to this comment?

**Comment by rdblue** [comment]
@hameizi, I'm mainly just waiting for a reply to the comment above. I wanted more information about that change. Thanks!

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java (hunk 2 above)
Comment (hameizi) [reply core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:124]: @rdblue yes,as you said that the logic here was intended to implement upsert just in case there are duplicate inserts without a delete. This logic can be retain, just i think duplicate inserts should be avoid by upstream semantic but not writer. If need i will rollback this change.

Comment (hameizi) [reply core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java]: @rdblue After test i think this is necessary, and i think old logic is error. Because old logic only write pos-delete in `write` function what should happen in `delete` function. And below test case is also puzzle:
(https://github.com/apache/iceberg/blob/9b6b5e0d2e760694e2abef73fa9036d1d8bbd014/data/src/test/java/org/apache/iceberg/io/TestTaskEqualityDeltaWriter.java#L324), this test write duplicate key but return just one data but user should avoid write duplicate key depend on config `write.upsert.enable` https://github.com/apache/iceberg/pull/2863. So this PR move pos-delete logic in write function to delete function.

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java [hunk 5]
```diff
@@ -52,8 +52,9 @@
   private final FileIO io;
   private final long targetFileSize;
 
-  protected BaseTaskWriter(PartitionSpec spec, FileFormat format, FileAppenderFactory<T> appenderFactory,
-                           OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
+  protected BaseTaskWriter(
+      PartitionSpec spec, FileFormat format, FileAppenderFactory<T> appenderFactory,
+      OutputFileFactory fileFactory, FileIO io, long targetFileSize) {
```
Comment (rdblue) [inline core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:57]: These lines don't need to change. Can you revert this?

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java (hunk 2 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:124]: Let's remove this change.

The logic here is to opportunistically catch duplicates when there are only inserts. This is not intended to replace the real upsert logic, which requires calling delete as you noted. Instead, it is here because we're updating the `insertedRowMap` and may get a previous insert location. When that happens, the right thing to do is to delete the duplicate row.

I also just realized that the changes below are incorrect. Instead of calling `internalPosDelete(key)`, this checks the `insertedRowMap` itself using `get`. The logic in internalPosDelete` used `remove` so that the entry was removed _before_ the insert occurred and we have a second opportunistic check.

To fix this, you should instead update `internalPosDelete` to return `true` if a row was deleted and `false` otherwise. Then you can update your `previous != null` check like this:

```java
    public void delete(T row) throws IOException {
      if (!internalPosDelete(structProjection.wrap(asStructLike(row)))) {
        eqDeleteWriter.write(row);
      }
    }
```

File: core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java (hunk 3 above)
Comment (rdblue) [reply core/src/main/java/org/apache/iceberg/io/BaseTaskWriter.java:157]: I noticed a problem with this: it doesn't remove the `PathOffset` from `insertedRowMap` the way `internalPosDelete` did. See my comment above.

**Comment by hameizi** [comment]
@rdblue fixed all. Can you help take a look?

File: flink/v1.13/flink/src/test/java/org/apache/iceberg/flink/sink/TestDeltaTaskWriter.java [hunk 6]
```diff
@@ -302,13 +302,13 @@ public void testPartitionedTableWithDataAndIdAsKey() throws IOException {
...
+    writer.write(createDelete(2, "aaa")); // 1 pos-delete.
 
     WriteResult result = writer.complete();
     Assert.assertEquals(1, result.dataFiles().length);
-    Assert.assertEquals(2, result.deleteFiles().length);
-    Assert.assertEquals(Sets.newHashSet(FileContent.EQUALITY_DELETES, FileContent.POSITION_DELETES),
-        Sets.newHashSet(result.deleteFiles()[0].content(), result.deleteFiles()[1].content()));
+    Assert.assertEquals(1, result.deleteFiles().length);
+    Assert.assertEquals(Sets.newHashSet(FileContent.POSITION_DELETES),
+            Sets.newHashSet(result.deleteFiles()[0].content()));
```
Comment (rdblue) [inline flink/v1.13/flink/src/test/java/org/apache/iceberg/flink/sink/TestDeltaTaskWriter.java:311]: Unnecessary whitespace change, and this is not the correct indentation.

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @hameizi! I merged this since there was only a style problem left.

**Comment by Xiangakun** [comment]
The comment in TestDeltaTaskWriter.testCdcEvents should also be updated since it may confuse the user.

image
